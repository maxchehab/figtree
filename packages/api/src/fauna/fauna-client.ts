import { Client, query as q, ExprVal } from 'faunadb';

import { FaunaEntity } from './interfaces/fauna-entity.interface';
import { LoginRequest } from '../common/interfaces/login-request.interface';
import { User } from '../common/interfaces/user.interface';

export class FaunaClient {
  private readonly client = new Client({
    secret: process.env.FAUNA_SECRET as string,
  });

  async userByLoginRequest(code: string): Promise<FaunaEntity<User> | null> {
    try {
      return this.client.query(
        q.Get(q.Match(q.Index('users_by_login_request_code'), code)),
      );
    } finally {
      return null;
    }
  }

  async updateUser(ref: ExprVal, data: Partial<User>) {
    this.client.query(q.Update(ref, { data }));
  }

  async userByToken(token: string): Promise<FaunaEntity<User> | null> {
    try {
      return this.client.query(
        q.Get(q.Match(q.Index('users_by_token'), token)),
      );
    } finally {
      return null;
    }
  }

  async createLoginRequest(
    code: string,
    createdAt: Date,
  ): Promise<FaunaEntity<LoginRequest>> {
    return this.client.query(
      q.Create(q.Collection('login_requests'), {
        data: { code, createdAt: createdAt.toISOString() },
      }),
    );
  }

  async loginRequestByCode(
    code: string,
  ): Promise<FaunaEntity<LoginRequest> | null> {
    try {
      return this.client.query(
        q.Get(q.Match(q.Index('login_requests_by_code'), code)),
      );
    } finally {
      return null;
    }
  }
}
