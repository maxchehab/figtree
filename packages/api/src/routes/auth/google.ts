import { google } from 'googleapis';

import { AuthStatus } from '../../common/interfaces/auth-status.enum';
import { FaunaClient } from '../../fauna/fauna-client';
import { HttpException } from '../../common/exceptions/http.exception';
import { User } from '../../common/interfaces/user.interface';
import lambda from '../../common/util/lambda.util';

const faunaClient = new FaunaClient();

const redirectURI =
  process.env.NODE_ENV === 'production'
    ? 'https://figtree.sh/auth/google'
    : 'http://localhost:3000/auth/google';

const oauth2Client = new google.auth.OAuth2(
  process.env.GOOGLE_CLIENT_ID,
  process.env.GOOGLE_CLIENT_SECRET,
  redirectURI,
);

export default lambda(async (req, res) => {
  const { code, state } = req.query;

  if (typeof code !== 'string') {
    throw new HttpException(400, 'Invalid code.');
  }

  if (typeof state !== 'string') {
    throw new HttpException(400, 'Invalid state.');
  }

  const { tokens } = await oauth2Client.getToken(code);
  oauth2Client.setCredentials(tokens);

  const googleClient = google.oauth2({
    auth: oauth2Client,
    version: 'v2',
  });

  const profile = await googleClient.userinfo.get();
  const { id, email } = profile.data;

  const data = {
    id,
    email,
    login_request_code: state,
    token: null,
    auth_status: AuthStatus.LoggingIn,
  } as Partial<User>;

  const user = await faunaClient.userByID(id as string);

  if (user) {
    await faunaClient.updateUser(user.ref, data);
  } else {
    await faunaClient.createUser(data);
  }

  return res.status(200).json({
    message: `You are logged in as ${email}. You may close this window.`,
  });
});
