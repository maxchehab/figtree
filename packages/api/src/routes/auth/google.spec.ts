import { oauth2_v2 } from 'googleapis';
import { OAuth2Client } from 'googleapis-common';
import axios from 'axios';
import http from 'http';
import listen from 'test-listen';

import { AuthStatus } from '../../common/interfaces/auth-status.enum';
import { FaunaClient } from '../../fauna/fauna-client';
import closeServer from '../../test/close-server';
import createTestLambda from '../../test/create-test-lambda';
import google from './google';

describe('google', () => {
  let server: http.Server;
  let url: string;

  beforeAll(async () => {
    server = createTestLambda(google);
    url = await listen(server);
  });

  afterAll(async () => {
    await closeServer(server);
  });

  describe('with invalid code', () => {
    it('returns 400, Invalid code.', async () => {
      const { status, data } = await axios.get(url, {
        validateStatus: () => true,
      });

      expect(status).toEqual(400);
      expect(data).toEqual({
        message: 'Invalid code.',
      });
    });
  });

  describe('with invalid state', () => {
    it('returns 400, Invalid code.', async () => {
      const { status, data } = await axios.get(url, {
        validateStatus: () => true,
        params: { code: 'valid code' },
      });

      expect(status).toEqual(400);
      expect(data).toEqual({
        message: 'Invalid state.',
      });
    });
  });

  describe('with valid code and state', () => {
    beforeAll(() => {
      jest
        .spyOn(OAuth2Client.prototype, 'getToken')
        .mockImplementation(async () => ({ tokens: {} }));

      jest
        .spyOn(oauth2_v2.Resource$Userinfo.prototype, 'get')
        .mockImplementation(async () => ({
          data: {
            id: 'usr_123',
            email: 'example@gmail.com',
          },
        }));
    });

    describe('with existing user', () => {
      let updateUser: any;

      beforeEach(() => {
        jest
          .spyOn(FaunaClient.prototype, 'userByID')
          .mockImplementationOnce(async () => ({
            ref: '@ref:123',
            data: {
              email: 'example@gmail.com',
              auth_status: AuthStatus.LoggedOut,
              id: 'usr_123',
              login_request_code: 'code_123',
              token: 'token_123',
            },
          }));

        updateUser = jest
          .spyOn(FaunaClient.prototype, 'updateUser')
          .mockImplementationOnce(async () => ({
            ref: '@ref:123',
            data: {
              email: 'example@gmail.com',
              auth_status: AuthStatus.LoggedIn,
              id: 'usr_123',
              login_request_code: 'code_123',
              token: 'token_123',
            },
          }));
      });

      it('updates user and returns 200', async () => {
        const { status, data } = await axios.get(url, {
          validateStatus: () => true,
          params: { code: 'code', state: 'state' },
        });

        expect(status).toEqual(200);
        expect(data).toEqual({
          message: `You are logged in as example@gmail.com. You may close this window.`,
        });

        expect(updateUser).toBeCalledTimes(1);
        expect(updateUser.mock.calls[0][1]).toMatchSnapshot();
      });
    });

    describe('with non-existing user', () => {
      let createUser: any;

      beforeEach(() => {
        jest
          .spyOn(FaunaClient.prototype, 'userByID')
          .mockImplementationOnce(async () => null);

        createUser = jest
          .spyOn(FaunaClient.prototype, 'createUser')
          .mockImplementationOnce(async () => ({
            ref: '@ref:123',
            data: {
              email: 'example@gmail.com',
              auth_status: AuthStatus.LoggedIn,
              id: 'usr_123',
              login_request_code: 'code_123',
              token: 'token_123',
            },
          }));
      });

      it('creates user and returns 200', async () => {
        const { status, data } = await axios.get(url, {
          validateStatus: () => true,
          params: { code: 'code', state: 'state' },
        });

        expect(status).toEqual(200);
        expect(data).toEqual({
          message: `You are logged in as example@gmail.com. You may close this window.`,
        });

        expect(createUser).toBeCalledTimes(1);
        expect(createUser.mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });
});
