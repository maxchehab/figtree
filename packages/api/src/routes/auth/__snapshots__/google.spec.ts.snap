// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`google with valid code and state with existing user updates user and returns 200 1`] = `
Object {
  "auth_status": "LOGGING_IN",
  "email": "example@gmail.com",
  "id": "usr_123",
  "login_request_code": "state",
  "token": null,
}
`;

exports[`google with valid code and state with non-existing user creates user and returns 200 1`] = `
Object {
  "auth_status": "LOGGING_IN",
  "email": "example@gmail.com",
  "id": "usr_123",
  "login_request_code": "state",
  "token": null,
}
`;
